fix #1793 Reject Context.of calls with duplicate keys

simonbasle opened the pull request.

bsideup commented on reactor-core/src/main/java/reactor/util/context/Context.java:
```diff
@@ -72,6 +94,7 @@ static Context of(Object key, Object value) {
 	 */
 	static Context of(Object key1, Object value1,
 			Object key2, Object value2) {
+		checkDuplicateKeys(key1, key2);
```
AFAIR this creates an allocation (an array of varargs), while the implementation is as simple as `if (key1.equals(key2))`.
Maybe we shouldn't use the generic check method here?

simonbasle replied:
good point, originally intended to limit the impact for the 2 k/v case (which is likely one of the most used along with 1 k/v and `empty()` cases) but moving that special case from `checkDuplicateKey` to the main method would save additional allocation.

robotmrv commented on reactor-core/src/main/java/reactor/util/context/Context.java:
```diff
@@ -89,6 +116,13 @@ static Context of(Object key1, Object value1,
 			Object key2, Object value2,
 			Object key3, Object value3) {
+		//prefer direct check for the simple 3 keys case over checkDuplicateKeys (saves up 1 array allocation)
+		if (key1.equals(key2) || key1.equals(key3)) {
```
there is no null check so it is possible to catch NPE.  
Intresting. ContextX constructors allow null keys and values but put() method does not

simonbasle replied:
mmh that is indeed a shortcoming, although the IDE should report this kind of usage (as part of a `@NonNullApi`). let me open a separate issue to enforce null checks in the constructors.

simonbasle replied:
see #1797 and PR #1800

bsideup commented on reactor-core/src/main/java/reactor/util/context/Context.java:
```diff
@@ -61,6 +62,28 @@ static Context of(Object key, Object value) {
+	 *
+	 * @param keys the keys to check for duplicates, by looping over the combinations
+	 */
+	static void checkDuplicateKeys(Object... keys) {
```
I just realized.... Since the method is defined in the interface, it will be public. I guess we don't want to leak it to public, don't we?

simonbasle replied:
ah snap you're right we don't want that to leak... since the approach here is to throw, we don't have to decide between several candidate implementations so we could move this logic in the constructors 🤔

bsideup replied:
you could also create an internal class in `Context.java` just for the validation

simonbasle replied:
I moved the static method to `Context4`, which is accessible by `Context5` (the only other usage of the method)

bsideup commented on reactor-core/src/main/java/reactor/util/context/Context.java:
```diff
@@ -61,6 +62,28 @@ static Context of(Object key, Object value) {
+		int size = keys.length;
+		if (size < 2) return;
+
+		for (int i = 0; i < keys.length - 1; i++) {
```
```suggestion
		for (int i = 0; i < size - 1; i++) {
```

bsideup commented on reactor-core/src/main/java/reactor/util/context/Context.java:
```diff
@@ -61,6 +62,28 @@ static Context of(Object key, Object value) {
+
+		for (int i = 0; i < keys.length - 1; i++) {
+			Object key = keys[i];
+			for (int j = i + 1; j < keys.length; j++) {
```
```suggestion
			for (int j = i + 1; j < size; j++) {
```

bsideup commented on reactor-core/src/main/java/reactor/util/context/Context.java:
```diff
@@ -61,6 +62,28 @@ static Context of(Object key, Object value) {
+	 */
+	static void checkDuplicateKeys(Object... keys) {
+		int size = keys.length;
+		if (size < 2) return;
```
`for (int i = 0; i < size - 1; i++) {` will ignore it anyways

bsideup commented on reactor-core/src/main/java/reactor/util/context/Context.java:
```diff
@@ -72,6 +95,10 @@ static Context of(Object key, Object value) {
 			Object key2, Object value2) {
+		//prefer direct check for the simple 2 keys case over checkDuplicateKeys (saves up 1 array allocation)
+		if (key1.equals(key2)) {
+			throw new IllegalArgumentException("Found duplicate key in Context.of() with 2 key-value pairs: " + key1);
```
Since the stracktrace will have `Context.of` already, maybe we should simplify the message to something like `"Duplicated key: " + key1`?

simonbasle replied:
I changed the message to the format `Key #keyN (valueN) is duplicated` across the board

bsideup commented on reactor-core/src/main/java/reactor/util/context/Context.java:
```diff
@@ -132,9 +167,42 @@ static Context of(Object key1, Object value1,
+		if (size == 0) return Context.empty();
+		if (size <= 5) {
+			Map.Entry[] entries = keyValuePairs.entrySet().toArray(new Map.Entry[size]);
+			if (size == 1) return new Context1(entries[0].getKey(), entries[0].getValue());
```
WDYT about using `switch` here?

simonbasle replied:
done

bsideup commented on reactor-core/src/main/java/reactor/util/context/Context.java:
```diff
@@ -132,9 +167,42 @@ static Context of(Object key1, Object value1,
+	 * @implNote this method compacts smaller maps into a relevant fields-based implementation
+	 * when number of key-value pairs is under 6.
+	 */
+	static Context of(Map<?, ?> keyValuePairs) {
```
the variable name is a bit misleading because there are no "pairs", map is a map :) Maybe just "values"?

simonbasle replied:
map is a container of key-value pairs, isn't it? but yeah, I should simply say `when size of the map is less than 6`.

simonbasle replied:
(my bad thought this referred to the javadoc comment... I'll also change the parameter name)

simonbasle commented:
@bsideup I'll rebase this on top of #1800 once merged and forwarded into `master`

simonbasle commented:
I have rebased on top of now merged #1800, so I have changed the last commit a bit: `checkDuplicateKeys` is now `checkKeys` and integrates the null check. We ensure the last key in the array is checked for null after the loops, since the outer loop never visits the last key.

robotmrv commented on reactor-core/src/main/java/reactor/util/context/ContextN.java:
```diff
@@ -29,42 +30,45 @@
 	}
 
 	ContextN(Map<Object, Object> sourceMap, Map<?, ?> other) {
 		super(sourceMap.size() + other.size(), 1f);
```
access to probably null value 
both bellow validations are useless - they will never work but just allocate message Strings

bsideup replied:
FYI constant strings are stored in the constant pool and are not allocated every time

simonbasle replied:
I'll follow up with a polish moving this part one step above

robotmrv commented on reactor-core/src/main/java/reactor/util/context/Context.java:
```diff
@@ -140,6 +141,45 @@ static Context of(Object key1, Object value1,
+	 * when map size is less than 6.
+	 */
+	static Context of(Map<?, ?> map) {
+		int size = map.size();
```
validation from the ContextN() constructor will never work.  
but it is still obvious NPE on the `map`. maybe it is ok)
